### ISSUE TITLE
AssertDriverContentsWithFrameAre - Adds spurious spaces at ends of lines

### ISSUE THREAD

**Issue by tig** [gui-cs/Terminal.Gui#2113]
Tests in `ViewTests.cs` are using `AssertDriverContentsWithFrameAre` and have been coded to account for the fact that it is adding spurious spaces to the "actual".

E.g. (note 4 extra spaces after "Edit")

<img width="421" alt="image" src="https://user-images.githubusercontent.com/585482/197064312-41be1623-1a59-4ec7-9581-944d2a9c9903.png">

If `AssertDriverContentsAre` is used, these spurious spaces do not appear in actual and are not needed in expected:

<img width="396" alt="image" src="https://user-images.githubusercontent.com/585482/197064635-049a6189-971b-4aef-94cb-3c26985e0372.png">

I just spent several hours debugging ViewTests trying to figure out this. Ugh.

**Comment by BDisp** [comment]
`AssertDriverContentsWithFrameAre` always returns a rectangle as a frame does.
`AssertDriverContentsAre` eliminate all spaces at the end, but sometime doesn't  full return what the `Contents` have.

If you have an output like this:
```cs
┌────────────────────────────┐
                              
│                            │
│                            │
│                            │
└────────────────────────────┘
```

`AssertDriverContentsWithFrameAre` will also output the same:
```cs
┌────────────────────────────┐
                              
│                            │
│                            │
│                            │
└────────────────────────────┘
```

`AssertDriverContentsAre` will output this:
```cs
┌────────────────────────────┐
│                            │
│                            │
│                            │
└────────────────────────────┘
```
That is, it will eliminate all spaces falsifying the actual exit. That why I created another method.

**Comment by tig** [comment]
I'm not convinced it works right.

However, even if it does, I don't like the idea of writing tests that rely on spaces that are not visible in the editor. This is especially problematic when debugging tests.

If you think this function is bug free I'll take your word for it.

**Comment by BDisp** [comment]
> If you think this function is bug free I'll take your word for it.

Yes, I made that way. It isn't a bug. It returns a Rect and the other one is void.

**Comment by tznind** [comment]
> I don't like the idea of writing tests that rely on spaces that are not visible in the editor. This is especially problematic when debugging tests.

I agree.  When I wrote `AssertDriverContentsAre`, I deliberately eliminated trailing whitespace on every line because of exactly this.  Especially since it is evaluating the full 'console' area not just the area that the View is rendered into.  I didn't want to get into adding 30 spaces on the end of each line of test string!.

>  AssertDriverContentsAre eliminate all spaces at the end, but sometime doesn't full return what the Contents have.
> If you have an output like this:
> [...]

Ah so `AssertDriverContentsAre` removes empty lines in the middle of the output? That is a bug then for sure.  It should only remove empty lines at the end.  I will try to find some time to fix that.

**Comment by BDisp** [comment]
That why I didn't make any change to the `AssertDriverContentsAre` because I was getting a lot of unit tests errors when I tried to change it and then I created a new method `AssertDriverContentsWithFrameAre` which return a `Rect`. The first works well with only a view, but to output various views at the same time it wasn't consistent with the `ConsoleDriver.Contents`. I don't have to write the expected output by hand. Normally when I have certain sure that the output is right I do Shift+F9 on VS2022 to show the quick watch and copy/paste the output to the expected value. Another feature I wanted is return a frame to validate the rectangle output dimensions. If removing all empty spaces is important then is better use the `AssertDriverContentsAre`, but please don't changes the current `AssertDriverContentsWithFrameAre` behavior, please :-)

**Comment by BDisp** [comment]
> However, even if it does, I don't like the idea of writing tests that rely on spaces that are not visible in the editor. This is especially problematic when debugging tests.

Well this isn't totally correct because they are in the `ConsoleDriver.Contents`. When a view is cleared a space character is added in each column and row. So, all the spaces all visible on the editor.

**Comment by tig** [comment]
Please don't misunderstand me. I really appreciate that we have so many, great, unit tests and I love the inventiveness that led to `AssertDriverContentsWithFrameAre`!

However, we now have a situation where the combination of:

1)  Code, such as `Menu.cs`, where UI elements have been massaged with `+ 3`, `"  "`, and other non-documented tweaks.

AND

2) Tests, which rely on typed whitespace that is not readily visible in the editor

has led to very fragile/brittle unit tests that with huge amounts of duplicated code.

Please take a look at the PR I did for `MenuTests.cs` to see how I consolidated the logic of creating "expected" strings. It's not perfect, but it makes it so that it is far easier to update ALL tests quickly if a minor change to the code (e.g. changing a `"  "` to a `" "`) to adjust the look & feel of something.

In the future, as we add new tests, we should try to reduce the amount of duplicated test code in a similar way.

**Comment by BDisp** [comment]
I'll will investigate if the output is a really the same as Contents output in each column and row. Thanks.

**Comment by tig** [comment]
I still think this API is wonky.

However, for new tests that don't absolutely need  to depend on `View.Redraw`, one can now use:

https://github.com/gui-cs/Terminal.Gui/blob/b4552ee14b0e9249e29c6baea06c40da13c27440/UnitTests/TestHelpers.cs#L316-L358

**Comment by BDisp** [comment]
> I still think this API is wonky.
> 
I created that method to check if the output cols and rows are the same as expected for views and believe me that I discovered many tests that didn't throws with the `AssertDriverContentsAre` and with the `AssertDriverContentsWithFrameAre` did throws, because the position on the screen wasn't the same. The spurious spaces at ends of lines is needed because the method returns a `Rect`.

> However, for new tests that don't absolutely need to depend on `View.Redraw`, one can now use:

I absolutely agree. I only will use the `AssertDriverContentsWithFrameAre` to test views output.

So, I appreciate you do not removed and may be we can close this issue. Thanks.

**Comment by tig** [comment]
Closing. Stale.
